### ISSUE TITLE
Use of `RawCreateParams` for `preprocess` context

### ISSUE THREAD

**Issue by ian-schu** [colinhacks/zod#1100]
Hey there 👋🏻 I'm coming in from `io-ts` and very happy to have found this project.

**TL;DR:** assuming I've set up some of my own-built schemas correctly, I can't seem to get custom errors out of a `preprocess`-based schema.

---

What I have is an own-rolled schema like the following, based on the `Dates` example from your docs:
```ts
export const dateString = (
  params?: RawCreateParams // <-- added params
) =>
  z.preprocess(
    (input: unknown) => {
      if (!(typeof input == 'string')) {
        return;
      }

      const date = new Date(input);

      if (isNaN(date.getTime())) {
        return;
      }
      return date;
    },
    z.date(),
    params
  );
```

Pretty basic stuff. The idea is to use it like so:

```ts
const someSchema = z.object({
  updated_at: dateString({
    invalid_type_error: "Field `updated_at` had wrong type"
  })
})
```

Problem is whenever I do this, the custom error message doesn't seem to survive. I've tried all the fields under `RawCreateParams`, but still just get the default error object when parsing fails.

By contrast, doing this with a built-in schema works just fine:

```ts
const someSchema = z.object({
  someString: z.string({
    invalid_type_error: 'Custom error message here!',
  }),
});
```

So, what am I doing wrong? Or is this a bug with `preprocess`?

**Comment by ian-schu** [comment]
Possibly related to https://github.com/colinhacks/zod/issues/696
